src/writeData/subscriptions/context/subscription.create.tsx

```diff
@@ -0,0 +1,149 @@
+import {getOrg} from 'src/organizations/selectors'
+import {sanitizeForm} from '../utils/form'
+
+// Types
+import {RemoteDataState} from 'src/types'
+import {Subscription} from 'src/types/subscriptions'
+
+export interface SubscriptionCreateContextType {
+  create: () => void
+  formContent: Subscription
+  formComplete: boolean
+  setFormComplete: (boolean) => void
+  updateForm: (formContent) => void
+  loading: RemoteDataState
+}
+
+export const DEFAULT_CONTEXT: SubscriptionCreateContextType = {
+  create: () => {},
+  formComplete: false,
+  formContent: {
+    name: '',
+    description: '',
+    protocol: 'MQTT',
+    brokerHost: '',
+    brokerPort: 0,
+    brokerUsername: '',
+    brokerPassword: '',
+    brokerCert: '',
+    brokerKey: '',
+    topic: '',
+    dataFormat: 'lineprotocol',
+    jsonMeasurementKey: {
+      name: '',
+      path: '',
+      type: 'string',
+    },
+    jsonFieldKeys: [
+      {
+        name: '',
+        path: '',
+        type: 'string',
+      },
+    ],
+    jsonTagKeys: [
+      {
+        name: '',
+        path: '',
+        type: 'string',
+      },
+    ],
+    jsonTimestamp: {
+      name: '',
+      path: '',
+      type: 'string',
+    },
+    stringMeasurement: {
+      pattern: '',
+      name: '',
+    },
+    stringFields: [
+      {
+        pattern: '',
+        name: '',
+      },
+    ],
+    stringTags: [
+      {
+        pattern: '',
+        name: '',
+      },
+    ],
+    stringTimestamp: {
+      pattern: '',
+      name: '',
+    },
+    bucket: 'nifi',
+    qos: 0,
+  },
+  setFormComplete: () => {},
+  updateForm: () => {},
+  loading: RemoteDataState.NotStarted,
+} as SubscriptionCreateContextType
+
+export const SubscriptionCreateContext = React.createContext<
+  SubscriptionCreateContextType
+>(DEFAULT_CONTEXT)
+
+export const SubscriptionCreateProvider: FC = ({children}) => {
+  const [formContent, setFormContent] = useState(DEFAULT_CONTEXT.formContent)
+  const [formComplete, setFormComplete] = useState(false)
+  const [loading, setLoading] = useState(RemoteDataState.Done)
+  const history = useHistory()
+  const org = useSelector(getOrg)
+  const create = (formContent?: Subscription): any => {
+    setLoading(RemoteDataState.Loading)
+    createAPI({data: formContent})
+      .then(() => {
+        setLoading(RemoteDataState.Done)
+        history.push(`/orgs/${org.id}/${LOAD_DATA}/${SUBSCRIPTIONS}`)
+        console.log('success')
+      })
+      .catch(() => {
+        setLoading(RemoteDataState.Error)
+        console.log('failure')
+      })
+  }
+
+  const updateForm = useCallback(
+    formContent => {
+      setFormContent(prev => ({
+        ...prev,
+        ...formContent,
+      }))
+    },
+    [formContent] // eslint-disable-line react-hooks/exhaustive-deps
+  )
+
+  useEffect(() => {
+    if (formComplete) {
+      sanitizeForm(formContent)
+      create(formContent)
+    }
+  }, [formComplete])
```

Comment:
instead of the `formComplete` boolean to denote when we should invoke `create`, could we export a function from the context? Something called `saveForm()` or something? And then the associated onClick would fire that function instead of setFormComplete. Or maybe I'm missing another use case for the `formComplete` flag - let me know.

Reply:
Currently theres no other use case than just triggering the create func at the end. I think it could be useful when we add something to check all the required fields are in there perhaps.

Reply:
swapped it out for a saveForm func